services/navigation/builtin/builtin_test.go

```diff
@@ -43,157 +52,294 @@ func setupNavigationServiceFromConfig(t *testing.T, configFilename string) (navi
 	test.That(t, err, test.ShouldBeNil)
 
+	// The motion config provides a single vision service
 	test.That(t, len(ns.(*builtIn).motionCfg.VisionServices), test.ShouldEqual, 1)
 }
 
 func TestStartWaypoint(t *testing.T) {
-	ctx := context.Background()
-	logger := golog.NewTestLogger(t)
+	t.Run("Calls motion.MoveOnGlobe for every waypoint", func(t *testing.T) {
+		callChan := make(chan moveOnGlobeReq)
+		injectMSensor := inject.NewMovementSensor("test_movement")
 
-	injectMS := inject.NewMotionService("test_motion")
-	cfg := resource.Config{
-		Name:  "test_base",
-		API:   base.API,
-		Frame: &referenceframe.LinkConfig{Geometry: &spatialmath.GeometryConfig{R: 100}},
-	}
+		injectMS := inject.NewMotionService("test_motion")
+		injectMS.MoveOnGlobeFunc = func(
+			ctx context.Context,
+			componentName resource.Name,
+			destination *geo.Point,
+			heading float64,
+			movementSensorName resource.Name,
+			obstacles []*spatialmath.GeoObstacle,
+			motionCfg *motion.MotionConfiguration,
+			extra map[string]interface{},
+		) (bool, error) {
+			callChan <- moveOnGlobeReq{
+				componentName:      componentName,
+				destination:        destination,
+				heading:            heading,
+				movementSensorName: movementSensorName,
+				obstacles:          obstacles,
+				motionCfg:          motionCfg,
+				extra:              extra,
+			}
+			return true, nil
+		}
+		ctx := context.Background()
+		logger := golog.NewTestLogger(t)
+		b := fakeBase(ctx, t, logger)
+		ns, teardown := setupNavService(ctx, t, injectMS, injectMSensor, b, logger)
+		defer teardown()
+
+		// add waypoint 1
+		pt1 := geo.NewPoint(1, 0)
+		err := ns.AddWaypoint(ctx, pt1, nil)
+		test.That(t, err, test.ShouldBeNil)
 
-	fakeBase, err := fakebase.NewBase(ctx, nil, cfg, logger)
-	test.That(t, err, test.ShouldBeNil)
+		// add waypoint 2
+		pt2 := geo.NewPoint(3, 1)
+		err = ns.AddWaypoint(ctx, pt2, nil)
+		test.That(t, err, test.ShouldBeNil)
 
-	fakeSlam := fakeslam.NewSLAM(slam.Named("foo"), logger)
-	limits, err := fakeSlam.Limits(ctx)
-	test.That(t, err, test.ShouldBeNil)
+		// test that added waypoints are returned by the waypoints function
+		wps, err := ns.Waypoints(ctx, nil)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, len(wps), test.ShouldEqual, 2)
+		wpSimilar(t, wps[0], pointToWP(pt1))
+		wpSimilar(t, wps[1], pointToWP(pt2))
 
-	localizer := motion.NewSLAMLocalizer(fakeSlam)
-	test.That(t, err, test.ShouldBeNil)
+		// set mode to Waypoint to begin MoveOnMap calls
+		err = ns.SetMode(ctx, navigation.ModeWaypoint, nil)
+		test.That(t, err, test.ShouldBeNil)
 
-	// cast fakeBase
-	fake, ok := fakeBase.(*fakebase.Base)
-	test.That(t, ok, test.ShouldBeTrue)
+		// confirm mode is now Waypoint
+		mode, err := ns.Mode(ctx, nil)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, mode, test.ShouldEqual, navigation.ModeWaypoint)
 
-	options := kinematicbase.NewKinematicBaseOptions()
-	options.PositionOnlyMode = false
+		expectedMotionCfg := &motion.MotionConfiguration{
```

Comment:
this also seems duplicative and can be moved outside the tests

Reply:
done